Client/WebFilters/ShieldStats/Adblock/CachedAdBlockEngine.swift

```diff
@@ -0,0 +1,149 @@
+// Copyright 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import BraveCore
+import Data
+import BraveShared
+
+/// An object that wraps around an `AdblockEngine` and caches some results
+/// and ensures information is always returned on the correct thread on the engine.
+public class CachedAdBlockEngine {
+  /// We cache the models so that they load faster when we need to poll information about the frame
+  private var cachedCosmeticFilterModels = FifoDict<URL, CosmeticFilterModel?>()
+  /// We cache the models so that they load faster when doing stats tracking or request blocking
+  private var cachedShouldBlockResult = FifoDict<String, Bool>()
+  /// We cache the user scripts so that they load faster on refreshes and back and forth
+  private var cachedFrameScriptTypes = FifoDict<URL, Set<UserScriptType>>()
+  
+  private var engine: AdblockEngine?
+  private let source: AdBlockEngineManager.Source
+  private let serialQueue: DispatchQueue
+  
+  init(engine: AdblockEngine, source: AdBlockEngineManager.Source, serialQueue: DispatchQueue) {
+    self.engine = engine
+    self.source = source
+    self.serialQueue = serialQueue
+  }
+  
+  deinit {
+    engine = nil
+  }
```

Comment:
This doesn't really make sense, the engine will be nil'd out anyways when the class deinit's anyways. If this is required, something else is up…

Reply:
I don't know what I was seeing but I removed this engine = nil and I see no leaks anymore.  Engines are deallocated just fine.